### PR TITLE
refactor `use_entities` and `ignore_entities`

### PR DESCRIPTION
**Pull request by chkoss**
Fixes issue #3923 

**Proposed changes**:
- Internally, instead of `use_entities` and `ignore_entities` only one property `used_entities` is used
- For dumping into yml, the internal property `used_entities` is transformed back into either `ignore_entities` or `use_entities`, whichever will contain fewer elements (or `use_entities: True` if all entities are in `used_entities`).

**Status (please check what you already did)**:
- [x] added some tests for the functionality
- [ ] updated the documentation
- [x] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

**Comment by ChristinaKoss** [comment]
Open questions:

- Do we care about the order of entities in `use_entities`? E.g. should it correspond to the order in which they are listed in `entities`?

- For the unit tests (not written yet): Is it better to add three new files to `data/test_domains` (one for each test: all entities contained in `used_entities`, more than half but not all, less than half) and load the domains from those files, or should I manually create domains in the respective test functions?

**Comment by ChristinaKoss** [comment]
Do we care about the order of entities in use_entities and ignore_entities? E.g. should it correspond to the order in which they are listed in entities? For now, they are just sorted alphabetically.

**Comment by wochinge** [comment]
seems a bunch of tests are failing @ChristinaKoss https://travis-ci.com/RasaHQ/rasa/jobs/282991188#L6706

File: rasa/core/domain.py [hunk 1]
```diff
@@ -265,30 +265,61 @@ def collect_slots(slot_dict: Dict[Text, Any]) -> List[Slot]:
...
                 raise InvalidDomain(
                     "Intents are not unique! Found two intents with name '{}'. "
                     "Either rename or remove one of them.".format(name)
                 )
 
+            if (
+                properties.get("used_entities") is None
+            ):  # important to differentiate between None and [] here
+
+                properties.setdefault("use_entities", True)
```
Comment (ricwo) [inline rasa/core/domain.py]: can we put all of this in one function that modifies the dict and returns what you called `include` below?

File: rasa/core/domain.py [hunk 2]
```diff
@@ -265,30 +265,61 @@ def collect_slots(slot_dict: Dict[Text, Any]) -> List[Slot]:
...
+                        f" excluded for intent '{name}'."
+                        f"Excluding takes precedence in this case. "
+                        f"Please resolve that ambiguity.",
+                        docs=DOCS_URL_DOMAINS
+                        + "#ignoring-entities-for-certain-intents",
+                    )
+
+                properties["used_entities"] = wanted_entities
+                properties.pop("use_entities", None)
+                properties.pop("ignore_entities")
```
Comment (ricwo) [inline rasa/core/domain.py]: I think all of this should be its own function as well

File: rasa/core/domain.py [hunk 3]
```diff
@@ -729,13 +742,30 @@ def as_dict(self) -> Dict[Text, Any]:
...
+        for idx, intent_info in enumerate(domain_data["intents"]):
+            for name, intent in intent_info.items():
+                use_entities = set(intent.pop("used_entities"))
+                ignore_entities = set(domain_data["entities"]) - use_entities
+                if len(use_entities) == len(domain_data["entities"]):
+                    intent["use_entities"] = True
+                elif len(use_entities) <= len(domain_data["entities"]) / 2:
+                    intent["use_entities"] = list(use_entities)
+                else:
+                    intent["ignore_entities"] = list(ignore_entities)
```
Comment (ricwo) [inline rasa/core/domain.py]: what if you want to use all entities and just exclude a single one? IMO something like `use_entities: True` and `ignore_entities: [entity_1]` should be possible

File: rasa/core/domain.py [hunk 4]
```diff
@@ -729,13 +742,30 @@ def as_dict(self) -> Dict[Text, Any]:
...
+        return domain_data
+
+    def cleaned_domain_for_file(self) -> Dict[Text, Any]:
```
Comment (ricwo) [inline rasa/core/domain.py]: do we have to rename this method? there's downstream code in rasa-x relying on this, so if we do change it we need to make sure it's changed there as well. this might lead to unnecessary incompatibilities

File: rasa/core/domain.py [hunk 5]
```diff
@@ -265,30 +265,61 @@ def collect_slots(slot_dict: Dict[Text, Any]) -> List[Slot]:
...
+
             if name in intent_properties.keys():
                 raise InvalidDomain(
                     "Intents are not unique! Found two intents with name '{}'. "
                     "Either rename or remove one of them.".format(name)
                 )
 
+            if (
+                properties.get("used_entities") is None
+            ):  # important to differentiate between None and [] here
```
[the remark this answers is not in the document]
Comment (ChristinaKoss) [reply rasa/core/domain.py]: One case is that the field exists and contains '[]' (no entities are used), the other case is that the field doesn't exist (because instead there is a field `use_entities` or `ignore_entities`).

But you are right, this allows adding `used_entities: None` to an intent in your `domain.yml`, which produces unwanted behaviour: It would use the default `use_entities: True` and put all entities into `used_entities`.

File: rasa/core/domain.py [hunk 6]
```diff
@@ -265,30 +265,61 @@ def collect_slots(slot_dict: Dict[Text, Any]) -> List[Slot]:
...
+            name, properties = list(intent.items())[0]
+
             if name in intent_properties.keys():
                 raise InvalidDomain(
                     "Intents are not unique! Found two intents with name '{}'. "
                     "Either rename or remove one of them.".format(name)
                 )
 
+            if (
+                properties.get("used_entities") is None
```
[the remark this answers is not in the document]
Comment (ChristinaKoss) [reply rasa/core/domain.py]: Not on purpose. We should't introduce a third option, `used_entities` should be internal only.

The reason why I added handling for an existing property `used_entities` was that otherwise `self.__class_from_dict(self.as_dict())` wouldn't work: `from_dict` calls the domain constructor, so `collect_intent_properties` in this case is called with `used_entities` existing instead of `use_entities` or `ignore_entities`.

A possible solution to this:
How about I change the `as_dict` function to do the transformation back from `used_entities` to `use_entities`/`ignore_entities`?
Tobias pointed out that `as_dict` is only used for things that should go "outside" anyways, so if we want `used_entities` to be internal only changing it like this would be better anyways. Right now, e.g. in `server.py::get_domain` we give back the internal version with `used_properties`, but probably we shouldn't.
What do you think?

File: rasa/core/domain.py [hunk 7]
```diff
@@ -265,30 +265,61 @@ def collect_slots(slot_dict: Dict[Text, Any]) -> List[Slot]:
...
-                    properties.setdefault("ignore_entities", [])
-                    if (
-                        properties["use_entities"] is None
-                        or properties["use_entities"] is False
-                    ):
-                        properties["use_entities"] = []
-            else:
-                name = intent
+            if not isinstance(intent, dict):
                 intent = {intent: {"use_entities": True, "ignore_entities": []}}
```
[the remark this answers is not in the document]
Comment (ChristinaKoss) [reply rasa/core/domain.py]: Not sure about this.

`intent` should be a dict with entries `use_entities` and `ignore_entities`. Before this line, `intent` either already is such a dict or is just a string, so this line simply "fixes" the intent variable if it was a string.

A reasonable alternative I could think of would be
`for incoming_intent in intents:
    if not isinstance(incoming_intent, dict):
        intent = {incoming_intent: {"use_entities": True, "ignore_entities": []}}
    else:
        intent = incoming_intent`

I would prefer to leave it as it is because I find it easier to read. It's a bit of a trade-off between simplicity and following conventions, I think. Do you think the alternative is better?

File: rasa/core/domain.py (hunk 3 above)
Comment (ChristinaKoss) [reply rasa/core/domain.py]: By default, all entities are included except the ones listed in `ignore_entities`. So the option you are asking for already exists.
A domain file with an intent that has `use_entities: True` and `ignore_entities: [entity_1]` is equivalent to one that only has `ignore_entities: [entity_1]` and no field `use_entities`. For writing back into a file, the "simplest" variant is chosen, so the unnecessary field is simply left out.

File: rasa/core/domain.py (hunk 4 above)
Comment (ChristinaKoss) [reply rasa/core/domain.py]: Thanks for pointing this out! I thought the new name is more descriptive, because "cleaned_domain" kind of implies that it is still a "domain" just cleaned up but here it has different properties after cleaning than an internally used domain. But the change isn't necessary, I'll change it back.

File: tests/core/test_domain.py [hunk 8]
```diff
@@ -616,17 +631,18 @@ def test_clean_domain():
             "utter_default": [{"text": "default message"}],
         },
         "actions": ["utter_default", "utter_goodbye", "utter_greet"],
+        "session_config": {
```
[the remark this answers is not in the document]
Comment (ChristinaKoss) [reply tests/core/test_domain.py]: Previously, this test only checked if the (hashes of) the domains resulting from the two dictionaries are the same. I think that doesn't achieve what it should. Say `actual` is the output of the cleaning function and `expected` is a properly cleaned domain. Even if the cleaning function did nothing, `assert hash(Domain.from_dict(actual))==hash(Domain.from_dict(expected))` would still pass.
So we have to directly compare the dictionaries to really test if the cleaning function properly cleans the domain.

Since a recent change in rasa, domains have a `session_config` property. Comparing the domains still worked because a default value is added when calling `Domain.from_dict(some_dict)`. For comparing the dicts directly, I had to manually add this property to the `expected` dict.

File: rasa/core/domain.py [hunk 9]
```diff
@@ -729,13 +742,30 @@ def as_dict(self) -> Dict[Text, Any]:
...
+                if len(use_entities) == len(domain_data["entities"]):
+                    intent["use_entities"] = True
+                elif len(use_entities) <= len(domain_data["entities"]) / 2:
+                    intent["use_entities"] = list(use_entities)
+                else:
+                    intent["ignore_entities"] = list(ignore_entities)
+        return domain_data
+
+    def cleaned_domain_for_file(self) -> Dict[Text, Any]:
+        """Fetch cleaned domain, replacing redundant keys with default values and replacing the internal
```
[the remark this answers is not in the document]
Comment (ChristinaKoss) [reply rasa/core/domain.py]: Good catch! I thought that `black` would take care of that, but I just tried and turns out that apparently it doesn't care about line length in docstrings.

File: rasa/core/domain.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ChristinaKoss) [reply rasa/core/domain.py]: This would also automatically be addressed/solved by what I suggest below (having `as_dict()` return something with property `use_entities` or `ignore_entities` instead of `used_entities`).

File: rasa/core/domain.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ChristinaKoss) [reply rasa/core/domain.py]: What do you mean? Making the names constants, e.g. `USE_ENTITIES_KEY="use_entities"`?

`use_entities` and `ignore_entities` should only be used in `domain.py`, `used_entities` is needed elsewhere. So we put them in different places according to where they are needed?

File: rasa/core/domain.py [hunk 10]
```diff
@@ -265,30 +265,61 @@ def collect_slots(slot_dict: Dict[Text, Any]) -> List[Slot]:
...
+                    properties["use_entities"] is None
+                    or properties["use_entities"] is False
+                ):
+                    properties["use_entities"] = []
+
+                # `use_entities` is either a list of explicitly included entities
+                # or `True` if all should be included
+                include = properties.get(
+                    "use_entities", True
+                )  # this should even work before setting defaults
```
[the remark this answers is not in the document]
Comment (ChristinaKoss) [reply rasa/core/domain.py]: re this and the following 3 comments: I took all your feedback on this code section and rewrote the section to make things more clear

File: rasa/core/domain.py (hunk 2 above)
Comment (ChristinaKoss) [reply rasa/core/domain.py]: re this and the previous comment: I'm not sure what functionalities you think are better in separate functions (i.e. starting from where to where). Can you clarify?

Comment (ricwo) [reply rasa/core/domain.py]: oh sorry, I thought I'd highlighted a whole block. I meant the whole modification of the `properties` dict, from `if properties.get("used_entities") is None` onwards. Don't worry if it ends up being much more code, this function is just getting a hard to follow

File: rasa/core/domain.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ricwo) [reply rasa/core/domain.py]: > What do you mean? Making the names constants, e.g. USE_ENTITIES_KEY="use_entities"?

exactly

yes, I'd move `use_entities` and `ignore_entities` to `core.constants.py`, and leave `used_entities`, as a constant, in `domain.py`

File: rasa/core/domain.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ricwo) [reply rasa/core/domain.py]: yes, `get_domain()` should not return `used_entities` as this is returned to the user. making the transformation back in `as_dict()` makes sense

File: rasa/core/domain.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ricwo) [reply rasa/core/domain.py]: Up to you, but I think your suggestion looks great 👍

File: rasa/core/domain.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (ricwo) [reply rasa/core/domain.py]: exactly, it doesn't wrap strings that are too long in general

File: tests/core/test_domain.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (ricwo) [reply tests/core/test_domain.py]: I think this test just tries to ensure that the format in `default_unfeaturized_entities.yml` is already properly cleaned, so that the `as_dict()` versions of `expected` and `default_unfeaturized_entities.yml` are the same (done by `domain.__hash__()`). both are fine, your test covers that implicitly 👍

File: rasa/core/domain.py (hunk 3 above)
Comment (ricwo) [reply rasa/core/domain.py]: 👍

**Comment by ricwo** [comment]
@ChristinaKoss please don't forget to add changelog entry for this

File: rasa/core/domain.py [hunk 11]
```diff
@@ -732,10 +753,37 @@ def persist(self, filename: Union[Text, Path]) -> None:
...
+            elif len(use_entities) <= len(entities) / 2:
+                intent_for_file["use_entities"] = list(use_entities)
+            else:
+                intent_for_file["ignore_entities"] = list(ignore_entities)
+            intent_properties_for_file.append({intent_name: intent_for_file})
+
+        return intent_properties_for_file
 
+    def cleaned_domain(self) -> Dict[Text, Any]:
+        """Fetch cleaned domain, replacing redundant keys with default values and
```
Comment (wochinge) [inline rasa/core/domain.py]: optional-ish: I think this is too much `what` and to less `why`. The docstring should describe `why` we are cleaning the domain.

File: rasa/core/domain.py [hunk 12]
```diff
@@ -264,31 +264,68 @@ def collect_slots(slot_dict: Dict[Text, Any]) -> List[Slot]:
...
+    ) -> Dict[Text, Any]:
+        """Transform the intent properties `use_entities` and `ignore_entities` that are
+        used in domain files into one internal property `used_entities`. Use default
+        values where the intent properties are not explicitly provided.
+        """
+        name, properties = list(intent.items())[0]
+
+        properties.setdefault("use_entities", True)
+        properties.setdefault("ignore_entities", [])
+        if properties["use_entities"] is None or properties["use_entities"] is False:
```
Comment (wochinge) [inline rasa/core/domain.py]: can be shortened to:
```suggestion
        if not properties["use_entities"]:
```

File: rasa/core/domain.py [hunk 13]
```diff
@@ -264,31 +264,68 @@ def collect_slots(slot_dict: Dict[Text, Any]) -> List[Slot]:
...
+        # explicitly.
+        explicitly_included = isinstance(properties["use_entities"], list)
+        ambiguous_entities = included_entities.intersection(excluded_entities)
+        if explicitly_included and ambiguous_entities:
+            raise_warning(
+                f"Entities: '{ambiguous_entities}' are explicitly included and"
+                f" excluded for intent '{name}'."
+                f"Excluding takes precedence in this case. "
+                f"Please resolve that ambiguity.",
+                docs=DOCS_URL_DOMAINS + "#ignoring-entities-for-certain-intents",
```
Comment (wochinge) [inline rasa/core/domain.py]:
```suggestion
                docs=f"{DOCS_URL_DOMAINS}#ignoring-entities-for-certain-intents",
```

File: rasa/core/domain.py (hunk 12 above)
Comment (ricwo) [reply rasa/core/domain.py]: in any case, a comment would be helpful

Comment (wochinge) [reply rasa/core/domain.py]: that doesn't really matter right? Worst case is that an empty list is replaced with an empty list

Comment (ricwo) [reply rasa/core/domain.py]: true 👍

File: rasa/core/domain.py [hunk 14]
```diff
@@ -264,31 +268,68 @@ def collect_slots(slot_dict: Dict[Text, Any]) -> List[Slot]:
...
+                f"Entities: '{ambiguous_entities}' are explicitly included and"
+                f" excluded for intent '{name}'."
+                f"Excluding takes precedence in this case. "
+                f"Please resolve that ambiguity.",
+                docs=f"{DOCS_URL_DOMAINS}#ignoring-entities-for-certain-intents",
+            )
+
+        properties[USED_ENTITIES_KEY] = used_entities
+        properties.pop(USE_ENTITIES_KEY)
+        properties.pop(IGNORE_ENTITIES_KEY)
```
Comment (ricwo) [inline rasa/core/domain.py]:
```suggestion
        del properties[IGNORE_ENTITIES_KEY]
```

File: rasa/core/domain.py [hunk 15]
```diff
@@ -264,31 +268,68 @@ def collect_slots(slot_dict: Dict[Text, Any]) -> List[Slot]:
...
-                intent = {intent: {"use_entities": True, "ignore_entities": []}}
+            if not isinstance(intent, dict):
+                intent = {intent: {USE_ENTITIES_KEY: True, IGNORE_ENTITIES_KEY: []}}
 
+            name = list(intent.keys())[0]
             if name in intent_properties.keys():
                 raise InvalidDomain(
-                    "Intents are not unique! Found two intents with name '{}'. "
-                    "Either rename or remove one of them.".format(name)
+                    f"Intents are not unique! Found two intents with name '{name}'. "
```
Comment (ricwo) [inline rasa/core/domain.py]:
```suggestion
                    f"Intents are not unique! Found multiple intents with name '{name}'. "
```
there may be more than two, right?

File: rasa/core/domain.py [hunk 16]
```diff
@@ -264,31 +268,68 @@ def collect_slots(slot_dict: Dict[Text, Any]) -> List[Slot]:
...
+                    f"Either rename or remove one of them."
```
Comment (ricwo) [inline rasa/core/domain.py]:
```suggestion
                    f"Either rename or remove the duplicate ones."
```

File: rasa/core/domain.py [hunk 17]
```diff
@@ -264,31 +268,68 @@ def collect_slots(slot_dict: Dict[Text, Any]) -> List[Slot]:
...
                 )
 
+            intent = cls._transform_intent_properties_for_internal_use(intent, entities)
+
             intent_properties.update(intent)
```
Comment (ricwo) [inline rasa/core/domain.py]:
```suggestion
            intent_properties.update(intent)
            
```

File: rasa/core/domain.py [hunk 18]
```diff
@@ -741,16 +764,40 @@ def persist(self, filename: Union[Text, Path]) -> None:
         domain_data = self.as_dict()
         utils.dump_obj_as_yaml_to_file(filename, domain_data)
 
-    def cleaned_domain(self) -> Dict[Text, Any]:
-        """Fetch cleaned domain, replacing redundant keys with default values."""
+    def _transform_intents_for_file(self) -> List[Union[Text, Dict[Text, Any]]]:
+        """Replace the internal `used_entities` property by `use_entities` or
```
Comment (ricwo) [inline rasa/core/domain.py]: same here, let's follow the standard format

File: rasa/core/domain.py [hunk 19]
```diff
@@ -741,16 +764,40 @@ def persist(self, filename: Union[Text, Path]) -> None:
...
+                intent_props[USE_ENTITIES_KEY] = list(use_entities)
+            else:
+                intent_props[IGNORE_ENTITIES_KEY] = list(ignore_entities)
+            intent_props.pop(USED_ENTITIES_KEY)
+            intents_for_file.append({intent_name: intent_props})
+
+        return intents_for_file
 
+    def cleaned_domain(self) -> Dict[Text, Any]:
+        """Fetch cleaned domain to display or write into a file.
```
Comment (ricwo) [inline rasa/core/domain.py]: and here, please

File: rasa/core/domain.py [hunk 20]
```diff
@@ -741,16 +764,40 @@ def persist(self, filename: Union[Text, Path]) -> None:
...
         domain_data = self.as_dict()
+
         for idx, intent_info in enumerate(domain_data["intents"]):
             for name, intent in intent_info.items():
-                if intent.get("use_entities") is True:
-                    intent.pop("use_entities")
-                if not intent.get("ignore_entities"):
-                    intent.pop("ignore_entities", None)
+                if intent.get(USE_ENTITIES_KEY) is True:
+                    intent.pop(USE_ENTITIES_KEY)
```
Comment (ricwo) [inline rasa/core/domain.py]:
```suggestion
                    del intent[USE_ENTITIES_KEY]
```

File: rasa/core/domain.py (hunk 15 above)
Comment (ricwo) [reply rasa/core/domain.py]: might it make sense to first collect all duplicate entities in the loop, then raise the exception afterwards if there are any? that way we can give out a list of which ones are duplicate

File: rasa/core/domain.py [hunk 21]
```diff
@@ -265,30 +265,61 @@ def collect_slots(slot_dict: Dict[Text, Any]) -> List[Slot]:
 
     @staticmethod
     def collect_intent_properties(
-        intents: List[Union[Text, Dict[Text, Any]]]
+        intents: List[Union[Text, Dict[Text, Any]]], entities: List[Text]
     ) -> Dict[Text, Dict[Text, Union[bool, List]]]:
```
[the remark this answers is not in the document]
Comment (ricwo) [reply rasa/core/domain.py]: a docstring would be great

File: rasa/core/domain.py [hunk 22]
```diff
@@ -264,40 +268,89 @@ def collect_slots(slot_dict: Dict[Text, Any]) -> List[Slot]:
...
+        return intent
+
+    @classmethod
     def collect_intent_properties(
-        intents: List[Union[Text, Dict[Text, Any]]]
+        cls, intents: List[Union[Text, Dict[Text, Any]]], entities: List[Text]
     ) -> Dict[Text, Dict[Text, Union[bool, List]]]:
+
         intent_properties = {}
+        duplicates = []
```
Comment (ricwo) [inline rasa/core/domain.py]:
```suggestion
        duplicates = set()
```

File: rasa/core/domain.py [hunk 23]
```diff
@@ -264,40 +268,89 @@ def collect_slots(slot_dict: Dict[Text, Any]) -> List[Slot]:
...
+            if not isinstance(intent, dict):
+                intent = {intent: {USE_ENTITIES_KEY: True, IGNORE_ENTITIES_KEY: []}}
 
+            name = list(intent.keys())[0]
             if name in intent_properties.keys():
-                raise InvalidDomain(
-                    "Intents are not unique! Found two intents with name '{}'. "
-                    "Either rename or remove one of them.".format(name)
-                )
+                duplicates.append(name)
```
Comment (ricwo) [inline rasa/core/domain.py]:
```suggestion
                duplicates.add(name)
```
